### PR TITLE
Generate indexes individually

### PR DESCRIPTION
**Pull request by kamranHaven**
Applied a short term solution to move out index updates from a transaction, so that in case of conflicts, other index creation entries are not impacted.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -7,4 +7,5 @@ coverage
 .sonarlint
 
 ./test/**/*.js
-./test/**/*.js.map
\ No newline at end of file
+./test/**/*.js.map
+.vscode/
```
Comment (rognorak) [inline .gitignore]: we merging .vscode ?

File: components/persistor/lib/knex/db.ts [hunk 2]
```diff
@@ -805,41 +820,52 @@ module.exports = function (PersistObjectTemplate) {
...
+                        
+                        }
                     }
-                });
-            });       
+                ));
+            }
+            try { 
+                await syncIndexesForHierarchy('delete', dbChanges);
+                await syncIndexesForHierarchy('add', dbChanges);
```
Comment (rognorak) [inline components/persistor/lib/knex/db.ts:851]: is having all 3 of the calls to `syncIndexesForHierarchy` in one try catch ok ?

File: components/persistor/lib/knex/db.ts [hunk 3]
```diff
@@ -805,41 +820,52 @@ module.exports = function (PersistObjectTemplate) {
 
                     name = name.toLowerCase();
                     if (operation === 'add') {
-                        table[type](columns, name);
+                        await knex.schema.table(tableName, function (table) {
+                            table[type](columns, name);
+                        }).catch((error) => {
```
Comment (rognorak) [inline components/persistor/lib/knex/db.ts]: await should be wrapped in a try catch, await X.catch is mixing 2 paradigms.

File: .gitignore (hunk 1 above)
Comment (kamranHaven) [reply .gitignore]: this is just .gitignore to ignore it

File: components/persistor/lib/knex/db.ts (hunk 2 above)
Comment (kamranHaven) [inline components/persistor/lib/knex/db.ts]: I have switched things around in old commit. Let me know if what I have now makes sense!

File: components/persistor/lib/knex/db.ts (hunk 3 above)
Comment (kamranHaven) [reply components/persistor/lib/knex/db.ts]: changed

File: components/persistor/lib/knex/db.ts [hunk 4]
```diff
@@ -791,55 +791,88 @@ module.exports = function (PersistObjectTemplate) {
...
-                            component: 'persistor',
-                            dbChanges: dbChanges,
-                            module: 'db',
-                            activity: 'synchronizeIndexes',
-                            error: error
-                        }, 'Unable to update index');
-                    }
-                });
-            });       
+            await syncIndexesForHierarchy('delete', dbChanges);
```
Comment (newparad1gm) [inline components/persistor/lib/knex/db.ts]: I'm wondering if these need to be done in order or can we do a Promise.all, especially now that we are removing them from a single transaction.

Comment (rognorak) [reply components/persistor/lib/knex/db.ts]: I syncIndexesForHierarchy doesnt throw anymore, then it can be in a promise.all

Comment (kamranHaven) [reply components/persistor/lib/knex/db.ts]: done

File: components/persistor/lib/knex/db.ts [hunk 5]
```diff
@@ -791,55 +791,89 @@ module.exports = function (PersistObjectTemplate) {
...
-                            dbChanges: dbChanges,
-                            module: 'db',
-                            activity: 'synchronizeIndexes',
-                            error: error
-                        }, 'Unable to update index');
+                        name = name.toLowerCase();
+                        if (operation === 'add') {
+                            try {
+                                await knex.schema.table(tableName, async function (table) {
+                                    await table[type](columns, name);
```
Comment (nsaenz7) [inline components/persistor/lib/knex/db.ts:842]: are we sure the `table`being used here (passed in the callback) from `knex.schema.table` is async? we should verify that

**Review by nsaenz7 — APPROVED** [review]

**Review by newparad1gm — APPROVED** [review]

**Pull request merged**